js/application.js

```diff
@@ -0,0 +1,180 @@
+    inputQtyMarker.setAttribute("type","integer");
+    inputQtyMarker.setAttribute("class", "form-control input-sm qtyvalue numberEntry");
+    inputQtyMarker.setAttribute("onfocusout","subtotalCalc()")
+    inputQtyMarker.value = 0;
+    document.getElementById("insertQty").appendChild(inputQtyMarker);
+
+    // This gives each row a 'clear' button
+    var clearBtn = document.createElement("input");
+    clearBtn.setAttribute("class", "btn btn-sm clearButton");
+    clearBtn.setAttribute("id", "clearBtn" + [i]);
+    clearBtn.setAttribute("type", "submit");
+    clearBtn.setAttribute("value", "Clear");
+    document.getElementById("clear").appendChild(clearBtn);
+
+    // This gives each row a subtotal calc
+    var subTotal = document.createElement("li");
+    subTotal.setAttribute("class", "subtotalline");
+    subTotal.textContent = 0;
+    subTotals.appendChild(subTotal);
+  };
+  document.getElementById("Qty").appendChild(qty);
+  document.getElementById("subtotal").appendChild(subTotals);
+
+};
+
+setup();
+
+
+var subtotalCalc = function(){
+  var numSubTotals = document.getElementsByClassName("subtotalline").length;
+  var calcedSubTotal = 0;
+  for (var i = 0; i < numSubTotals; i++) {
+
+    var parsedPrice = document.getElementsByClassName("priceitem")[i].textContent.replace(/\D/g,'');
+    parsedPrice = parseFloat(parsedPrice);
+
+    calcedSubTotal = parseFloat(document.getElementsByClassName("qtyvalue")[i].value)* parsedPrice;
+    document.getElementsByClassName("subtotalline")[i].textContent = calcedSubTotal;
+  };
+};
+
+// Add New Item
+document.getElementById("addNewButton").addEventListener("click", function(){
+    // console.log("clicked")
+    var newItemEntry = document.getElementById("newItemEntry").value
+    var newPriceEntry = document.getElementById("newPriceEntry").value
+
+    if( newItemEntry == "" || newPriceEntry == "" ){
+      alert("Missing an entry grandma!");
+    }
+    else{
+      var newItemArray = [];
+      newItemArray.push(newItemEntry);
+      newItemArray.push(newPriceEntry);
+      shoppingList.push(newItemArray);
+      setup();
+
+    };
+  });
+
+// Calculate Total
+document.getElementById("calcTotalButton").addEventListener("click", function(){
+  // get length of all subtotal classes (it returns array) for each one take the .textContent
+  console.log("clicked")
+  var numSubTotals = document.getElementsByClassName("subtotalline").length;
+  var total = 0;
+  for (var i = 0; i < numSubTotals; i++) {
+    total += parseFloat(document.getElementsByClassName("subtotalline")[i].textContent);
+  };
+  document.getElementById("totalhere").textContent = total;
+});
+
+// Input Validation
+$(document).on('focusout', '.numberEntry', function(){
+  if (isNaN(this.value)){
+    alert("Entry must be a number grandma!");
+    this.value = 0;
+  }
+  else if(this.value == "" && this.id !== "newPriceEntry"){
+    this.value = 0;
+  };
+});
+
+$(document).on('focusout', '.textEntry', function(){
```

Comment:
You are the only one so far that have input validation on the name of the item! Awesome!